Add basic search functionality

GreenMan36 opened the pull request:
It's pretty scuffed as its been a while since I worked with React.
I added the `react-router-dom` for the `BrowserRouter` so I could use `useSearchParams` and `useNavigate`.

I'd like to move more of the state to the url, this is great as it allows you to share your search queries and what snippet you have open. However, that would be a major refactor, so I'll leave the idea here.

It's been like a year but [WDS](https://www.youtube.com/watch?v=oZZEI23Ri6E) had a video on it and I [Theo](https://www.youtube.com/watch?v=t3FUkq7yoCw&pp=ygUMc3RhdGUgaW4gdXJs) as well. I think, I didn't re-watch to check.

![image](https://github.com/user-attachments/assets/eba7b318-3a6a-479d-983f-8202cbf7259a)

GreenMan36 commented:
I don't have much experience with react so my architecture and code is likely not so clean.

Mathys-Gasnier commented:
I agree that moving the state to the URL should be something we work on, as it also allows to go back and forth in your history of visited pages

Mathys-Gasnier commented on src/main.tsx:
```diff
@@ -3,11 +3,14 @@ import { createRoot } from "react-dom/client";
-    <AppProvider>
-      <App />
-    </AppProvider>
+    <BrowserRouter>
```
Could you move the BrowserRouter to `App.tsx` just wrapping the routes ?

GreenMan36 replied:
Moved it, however the header needs to be wrapped as I use the route hook from the searchbar.

Mathys-Gasnier commented on src/components/SearchSnippetList.tsx:
Instead of having both `SearchSnippetList` and `SnippetList` i think you could merge the two together, just disabling the search path when not searching.

Mathys-Gasnier commented:
As you said you aren't really familiar with react router and all, if you want I can contribute to this PR improving the way it's used and handled

GreenMan36 commented:
> As you said you aren't really familiar with react router and all, if you want I can contribute to this PR improving the way it's used and handled

I'd really appreciate that! I could figure it out, but learning React Router while working on a public repo sounds like more effort than I want to put in right now.

Mathys-Gasnier commented on src/components/SnippetList.tsx:
```diff
@@ -41,8 +44,10 @@ const SnippetList = () => {
               </div>
-
               <h3 className="snippet__title">{snippet.title}</h3>
+              {query && (
```
IMO we could always display the description your opinion on this @dostonnabotov ?

GreenMan36 replied:
I actually had this when I made the search page!
It needs a bit more CSS to make the grid stay pretty though. Or the description needs to be cut off.

Otherwise some cards get bigger than others, causing whitespace underneath smaller cards.

Mathys-Gasnier replied:
I think it's a good idea to do it.

psychlone77 replied:
Yes, seems like a good idea. Try this to clamp the description to some amount of lines.
```
  .description {
      display: -webkit-box;
      -webkit-box-orient: vertical;
      -webkit-line-clamp: 2;
      overflow: hidden;
      text-overflow: ellipsis;
  }
```

GreenMan36 replied:
Might look weird on smaller screens.
![image](https://github.com/user-attachments/assets/f2afec14-eec5-40af-b443-b45ede31bbff)

Mathys-Gasnier commented on .github/ISSUE_TEMPLATE/ADD-LANGUAGE.yml:
What's this file ?

GreenMan36 replied:
That's not supposed to be there, I was messing around with GH templates like two hours ago, not sure how it even added an empty file, especially since I haven't touched it since.

Mathys-Gasnier reviewed (approved):
For me it looks good, let's wait on another review since it's a pretty big change

kuronekony4n commented:
![image](https://github.com/user-attachments/assets/2450a1fd-f55b-478c-b6cc-e1ff754c5a86)
you guys forget to change this thing

saminjay commented on src/hooks/useSnippets.ts:
```diff
@@ -14,8 +14,10 @@ export const useSnippets = () => {
 
-  const fetchedSnippets = data
-    ? data.find((item) => item.categoryName === category)?.snippets
+  const fetchedSnippets: SnippetType[] = data
```
instead of adding nested ternary operator, it would be better to populate it through a function.

GreenMan36 replied:
I did the thing 🫡

saminjay commented on src/components/SearchInput.tsx:
```diff
@@ -1,6 +1,31 @@
+    debounce((query: string) => {
+      if (query) {
+        setSearchParams({ q: query });
+        navigate(`/search?q=${encodeURIComponent(query.trim().toLowerCase())}`);
```
I don't think navigating urls every debounced request is a good thing. I know it is debounced but still sometimes people pause to think while searching, and it will clutter the history.

You can maybe search on `Enter`, or show results while typing but navigate the URL on Enter or onBlur

Mathys-Gasnier replied:
It's an SPA so it doesn't "refresh" the page, but yeah it will clutter the history, what you can do is use the replace feature of navigate [here](https://api.reactrouter.com/v7/interfaces/react_router.NavigateOptions.html#replace)

GreenMan36 replied:
I think on Enter and onBlur is a great idea! I'll look into it.

saminjay commented on src/App.tsx:
```diff
@@ -1,29 +1,24 @@
-import { useAppContext } from "./contexts/AppContext";
-
+import { Routes, Route, BrowserRouter } from "react-router-dom";
```
`react-router-dom` is not used anywhere in the project, it was just added in `package.json`. So I think, it will be better to use the latest version.

GreenMan36 replied:
I'm not sure where you're seeing it being added to the package.json. `react-router-dom` was part of the initial commit.
If you wish to have the dependencies updated, I recommend making a new PR for it.

saminjay replied:
It was added in the initial commie, but its not being used anywhere. Its just there.
Since you are using it in this PR, you can just make it like its being added for the first time and update it to the latest version.

Mathys-Gasnier commented on src/components/SearchInput.tsx:
```diff
@@ -1,58 +1,64 @@
+        // Remove search params and navigate to home if query is empty
         setSearchParams({});
-        navigate("/");
+        navigate("/", { replace: true });
```
This one shouldn't replace

Mathys-Gasnier commented on src/components/SearchInput.tsx:
```diff
@@ -1,58 +1,64 @@
+      // Only navigate if we're not already on the search page
+      if (location.pathname !== "/search") {
+        navigate("/search", {
+          replace: isCompletedSearch || location.pathname === "/search",
```
This one should always be false

GreenMan36 replied:
I'll look at it next year, I've been going crazy for the past four hours trying to get this all to behave.

Mathys-Gasnier reviewed (commented):
Apart from the two little comments i made, it seems about right

Mathys-Gasnier commented:
Oh and there is a weird flicker when you go from the search page to the hompage...

I think one way to make this easier would be to remove the separation between the search and hompage...
At least from the path, and only keep it as a query param, like if `?q=thing` is present then we search, else we dont.

GreenMan36 commented:
> Oh and there is a weird flicker when you go from the search page to the hompage...
> 
> I think one way to make this easier would be to remove the separation between the search and hompage...
> At least from the path, and only keep it as a query param, like if `?q=thing` is present then we search, else we dont.

I originally intended to make a different design, in my latest commit message I also conclude the same as you.

GreenMan36 commented:
> When using quicksnip, I find that this is the feature I am missing most. I'm not sure about the current state of development on this PR, but I wanted to help a little, so I finished the search feature based on the work already completed and feedback within this task (addressing some of the comments and smaller errors also).
> 
> @GreenMan36 If you are currently working on this and are nearing a solution please feel free to ignore, as a lot of my work is based on yours anyway. You are more than welcome to have a look and use any code from the patch file I will attach to this comment if you find it helpful.
> 
> Here is a [demo](https://jam.dev/c/dacd3887-9b53-4fdc-ad8a-8f66ffffbaa9). I'm also attaching the patch file so the work can be applied to this branch. I started from a more recent version of the `main` branch (so it might be better to apply it after pulling latest), as a lot of new changes have been added since work on this started. Thanks all.
> 
> [changes.patch](https://github.com/user-attachments/files/18301152/changes.patch)

Awesome work! I just shut down my PC but I guess imma boot it back up haha! Give me a few minutes!

I didn't share in the public channels but; https://lucaong.github.io/minisearch/

This might be a neat project to integrate.

GreenMan36 commented:
Yikes, I merged main to mine to be up-to-date and due to all the commit hooks I had to run lint and format but that seems to have changed a couple of things?

barrymun commented:
minisearch might be a nince addition in the future once we get the basic search up and running for sure.

GreenMan36 commented:
Yeah I'm having major issues with Prettier, ESlint and GIT all messing with formatting and line styles and whatnot. I think this branch and all my other are now broken.

barrymun commented:
> Yeah I'm having major issues with Prettier, ESlint and GIT all messing with formatting and line styles and whatnot. I think this branch and all my other are now broken.

Maybe the easiest thing to do might be to keep this PR active for now, create a new branch from latest and get everything working, apply the patch from above then make any changes you need from this branch? Maybe create a new PR as well if you need to more easily compare the changes.

GreenMan36 commented:
I did a hard reset of my branch on main, and its still broken. I recommend you make a new branch with your changes, make a new PR and you can mention this PR.

Your changes look good to me.

barrymun commented:
@GreenMan36 The new PR is here #159. Please feel free to leave a review, or suggest/make any changes that you think I've missed. Thanks

saminjay commented:
Closing this PR.
As a new one is already raised and author said its broken now.
